### ISSUE TITLE
When Patroni API  is unavailable no failover happens (maybe it should?)

### ISSUE THREAD

**Issue by 107142** [zalando/patroni#2263]
**Describe the bug**
I have a fairly standard Patroni setup (4 db backends and HAproxy cluster taking care of sending data to current master, etcd) somewhat similar to the one in extras.

Yesterday the Patroni API endpoint of a single (master) node stopped responding (but db backend itself was fine, streaming and responding to queries).
Without the API HAProxy had no way to know where the master is and declared the backend unavailable (as it should).
The reason for the API not responding is not known to me, there was nothing in the logs, it could also be some anomaly related to the OS. Manual switchover was not possible as /patroni endpoint did not respond, restarting the master node however caused fail-over.

My concern is mostly whether a node API should be checked and if, for whatever reason does not respond, declared that node is unhealthy (and if master trigger fail-over). Or perhaps there is a better way to determine current master?

**To Reproduce**
Firewall the Patroni API port during regular operation

**Expected behavior**
Fail-over should be triggered. Or perhaps having a setting whether unavailable API should be considered as fatal.

**Screenshots**

**Environment**
- Patroni version: 2.1.2
- PostgreSQL version: 13.5
- DCS (and its version): etcd 3.5.2

**Comment by CyberDem0n** [comment]
The node continues to run as a leader since it can update the leader key in etcd.
The REST API health check is used for traffic routing by HAProxy, but also REST API port must be accessible from all patroni nodes in this cluster.

We never seen so far the rest api failing on its own.
Without debug logs there is nothing I can help with.

**Comment by 107142** [comment]
I absolutely realize this is a probably an edge case. However it seems to me like absence of Patroni API could under right conditions cause database being unavailable and while finding why that happened is important for obvious reasons it seems to me like it would make sense to consider node unhealthy when API is unavailable.

But perhaps I'm missing something why this would be unwise?

Relevant portion of the logs. Server is running only Patroni (and Postgres), no other services running.
HAP:
```
Mar 31 00:03:43 haproxy1 haproxy/haproxy[1786]: 147.251.6.137:43408 [30/Mar/2022:22:03:43.403] pgsql-master pgsql-master/pgsql-node1 1/0/+0 +0 -- 66/51/51/51/0 0/0
Mar 31 00:03:44 haproxy1 haproxy/haproxy[1786]: [WARNING]  (10) : Server pgsql-master/pgsql-node1 is DOWN, reason: Layer4 connection problem, info: "Connection refused", check d
uration: 0ms. 0 active and 0 backup servers left. 50 sessions active, 0 requeued, 0 remaining in queue.
Mar 31 00:03:44 haproxy1 haproxy/haproxy[1786]: [ALERT]    (10) : proxy 'pgsql-master' has no server available!
Mar 31 00:03:44 haproxy1 haproxy/haproxy[1786]: Server pgsql-master/pgsql-node1 is DOWN, reason: Layer4 connection problem, info: "Connection refused", check duration: 0ms. 0 ac
tive and 0 backup servers left. 50 sessions active, 0 requeued, 0 remaining in queue.
Mar 31 00:03:44 haproxy1 haproxy/haproxy[1786]: proxy pgsql-master has no server available!
Mar 31 00:03:44 haproxy1 haproxy/haproxy[1786]: 147.251.6.109:34096 [30/Mar/2022:22:03:44.883] pgsql-master pgsql-master/<NOSRV> -1/-1/+0 +0 SC 16/1/0/0/0 0/0

```
Master at-the-time:
```
Mar 31 00:03:38 pgsql-node1 pgsql/pgsql[21836]: 2022-03-30 22:03:38,191 INFO: no action. I am (pgsql-node1.tld), the leader with the lock
Mar 31 00:03:40 pgsql-node1 pgsql/pgsql[21836]: 2022-03-30 22:03:40,145 INFO: no action. I am (pgsql-node1.tld), the leader with the lock
Mar 31 00:03:42 pgsql-node1 pgsql/pgsql[21836]: 2022-03-30 22:03:42,136 INFO: no action. I am (pgsql-node1.tld), the leader with the lock
Mar 31 00:03:43 pgsql-node1 pgsql/pgsql[21836]: 2022-03-31 00:03:43 CEST, DB=[unknown], USR=[unknown], CLIENT=fd00:dead:beef::16, APP=[unknown],LOG:  connection received: host=fd00:dead:beef::16 port=41050
Mar 31 00:03:43 pgsql-node1 pgsql/pgsql[21836]: 2022-03-31 00:03:43 CEST, DB=redacted, USR=redacted, CLIENT=fd00:dead:beef::16, APP=[unknown],LOG:  connection authorized: user=redacted database=redacted SSL enabled (protocol=TLSv1.3, cipher=TLS_AES_256_GCM_SHA384, bits=256, compression=off)
Mar 31 00:03:43 pgsql-node1 pgsql/pgsql[21836]: 2022-03-31 00:03:43 CEST, DB=[unknown], USR=[unknown], CLIENT=fd00:dead:beef::16, APP=[unknown],LOG:  connection received: host=fd00:dead:beef::16 port=41052
Mar 31 00:03:43 pgsql-node1 pgsql/pgsql[21836]: 2022-03-31 00:03:43 CEST, DB=redacted, USR=redacted, CLIENT=fd00:dead:beef::16, APP=[unknown],LOG:  connection authorized: user=redacted database=redacted SSL enabled (protocol=TLSv1.3, cipher=TLS_AES_256_GCM_SHA384, bits=256, compression=off)
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]: 2022-03-30 22:03:44,051 INFO: No local configuration items changed.
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]: 2022-03-30 22:03:44,074 ERROR: Couldn't start a service on ':::8008', please check your `restapi.listen` configuration
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]: 2022-03-30 22:03:44,074 ERROR: Failed to reload config_file=/etc/patroni/patroni.yml
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]: Traceback (most recent call last):
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]:   File "/usr/lib/python3/dist-packages/patroni/__init__.py", line 79, in reload_config
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]:     self.api.reload_config(self.config['restapi'])
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]:   File "/usr/lib/python3/dist-packages/patroni/api.py", line 871, in reload_config
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]:     self.__initialize(config['listen'], ssl_options)
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]:   File "/usr/lib/python3/dist-packages/patroni/api.py", line 776, in __initialize
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]:     self.__httpserver_init(host, port)
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]:   File "/usr/lib/python3/dist-packages/patroni/api.py", line 755, in __httpserver_init
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]:     HTTPServer.__init__(self, info[0][-1][:2], RestApiHandler)
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]:   File "/usr/lib/python3.9/socketserver.py", line 452, in __init__
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]:     self.server_bind()
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]:   File "/usr/lib/python3.9/http/server.py", line 138, in server_bind
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]:     socketserver.TCPServer.server_bind(self)
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]:   File "/usr/lib/python3.9/socketserver.py", line 466, in server_bind
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]:     self.socket.bind(self.server_address)
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]: OSError: [Errno 98] Address already in use
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]: 2022-03-30 22:03:44,240 INFO: no action. I am (pgsql-node1.tld), the leader with the lock
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]: 2022-03-31 00:03:44 CEST, DB=redacted, USR=redacted, CLIENT=fd00:dead:beef::16, APP=[unknown],LOG:  could not receive data from client: Connection reset by peer
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]: 2022-03-31 00:03:44 CEST, DB=redacted, USR=redacted, CLIENT=fd00:dead:beef::16, APP=[unknown],LOG:  could not receive data from client: Connection reset by peer
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]: 2022-03-31 00:03:44 CEST, DB=redacted, USR=redacted, CLIENT=fd00:dead:beef::16, APP=[unknown],LOG:  could not receive data from client: Connection reset by peer
Mar 31 00:03:44 pgsql-node1 pgsql/pgsql[21836]: 2022-03-31 00:03:44 CEST, DB=redacted, USR=redacted, CLIENT=fd00:dead:beef::16, APP=[unknown],LOG:  could not receive data from client: Connection reset by peer

```

Replica:

At 2022-03-30 22:20:06 I restarted the at-the-time master. Attempt to use patronictl failover (or switchover) failed on connection refused.

```
Mar 31 00:20:02 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:02 CEST, DB=[unknown], USR=[unknown], CLIENT=[local], APP=[unknown],LOG:  connection received: host=[local]
Mar 31 00:20:02 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:02 CEST, DB=redacted, USR=postgres, CLIENT=[local], APP=[unknown],LOG:  connection authorized: user=postgres database=redacted application_name=pg_dump
Mar 31 00:20:02 pgsql-node3 pgsql/pgsql[1338]: 2022-03-30 22:20:02,089 INFO: no action. I am (pgsql-node3.tld), a secondary, and following a leader (pgsql-node1.tld)
Mar 31 00:20:03 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:03 CEST, DB=[unknown], USR=[unknown], CLIENT=[local], APP=[unknown],LOG:  connection received: host=[local]
Mar 31 00:20:03 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:03 CEST, DB=redacted, USR=postgres, CLIENT=[local], APP=[unknown],LOG:  connection authorized: user=postgres database=redacted application_name=pg_dump
Mar 31 00:20:04 pgsql-node3 pgsql/pgsql[1338]: 2022-03-30 22:20:04,189 INFO: no action. I am (pgsql-node3.tld), a secondary, and following a leader (pgsql-node1.tld)
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-30 22:20:06,088 INFO: no action. I am (pgsql-node3.tld), a secondary, and following a leader (pgsql-node1.tld)
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-30 22:20:06,408 WARNING: Request failed to pgsql-node1.tld: GET https://pgsql-node1.tld:8008/patroni (HTTPSConnectionPool(host='pgsql-node1.tld', port=8008): Max retries exceeded with url: /patroni (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f5356732100>: Failed to establish a new connection: [Errno 111] Connection refused')))
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=[unknown], USR=[unknown], CLIENT=::1, APP=[unknown],LOG:  connection received: host=::1 port=35174
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=[unknown], USR=replicator, CLIENT=::1, APP=[unknown],LOG:  replication connection authorized: user=replicator SSL enabled (protocol=TLSv1.3, cipher=TLS_AES_256_GCM_SHA384, bits=256, compression=off)
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=[unknown], USR=[unknown], CLIENT=fd00:dead:beef::21, APP=[unknown],LOG:  connection received:host=fd00:dead:beef::21 port=35364
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=[unknown], USR=[unknown], CLIENT=fd00:dead:beef::23, APP=[unknown],LOG:  connection received:host=fd00:dead:beef::23 port=54066
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-30 22:20:06,672 INFO: promoted self to leader by acquiring session lock
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=[unknown], USR=replicator, CLIENT=fd00:dead:beef::21, APP=[unknown],LOG:  replication connection authorized: user=replicator application_name=pgsql-node2.tld SSL enabled (protocol=TLSv1.3, cipher=TLS_AES_256_GCM_SHA384, bits=256, compression=off)
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=[unknown], USR=replicator, CLIENT=fd00:dead:beef::21, APP=pgsql-node2.tld,ERROR:  replication slot "pgsql_node_2_tld" does not exist
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=[unknown], USR=replicator, CLIENT=fd00:dead:beef::21, APP=pgsql-node2.tld,STATEMENT:  START_REPLICATION SLOT "pgsql_node_2_tld" 249/CF000000 TIMELINE 69
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=[unknown], USR=replicator, CLIENT=fd00:dead:beef::23, APP=[unknown],LOG:  replication connection authorized: user=replicator application_name=pgsql_node_4_tld SSL enabled (protocol=TLSv1.3, cipher=TLS_AES_256_GCM_SHA384, bits=256, compression=off)
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=[unknown], USR=replicator, CLIENT=fd00:dead:beef::23, APP=pgsql-node4.tld,ERROR:  replication slot "pgsql_node_4_tld" does not exist
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=[unknown], USR=replicator, CLIENT=fd00:dead:beef::23, APP=pgsql-node4.tld,STATEMENT:  START_REPLICATION SLOT "pgsql_node_4_tld" 249/CF000000 TIMELINE 69
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-30 22:20:06,678 INFO: Lock owner: pgsql-node3.tld; I am pgsql-node3.tld
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-30 22:20:06,727 INFO: updated leader lock during promote
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: server promoting
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=, USR=, CLIENT=, APP=,LOG:  received promote request
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=, USR=, CLIENT=, APP=,FATAL:  terminating walreceiver process due to administrator command
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-30 22:20:06,753 INFO: cleared rewind state after becoming the leader
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=, USR=, CLIENT=, APP=,LOG:  redo done at 249/CFC08098
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=, USR=, CLIENT=, APP=,LOG:  last completed transaction was at log time 2022-03-30 22:04:14.140038+00
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=, USR=, CLIENT=, APP=,LOG:  selected new timeline ID: 70
Mar 31 00:20:06 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:06 CEST, DB=, USR=, CLIENT=, APP=,LOG:  archive recovery complete
Mar 31 00:20:07 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:07 CEST, DB=, USR=, CLIENT=, APP=,LOG:  database system is ready to accept connections
Mar 31 00:20:07 pgsql-node3 pgsql/pgsql[1338]: 2022-03-31 00:20:07 CEST, DB=[unknown], USR=[unknown], CLIENT=fd00:dead:beef::16, APP=[unknown],LOG:  connection received:host=fd00:dead:beef::16 port=53580

```

**Comment by CyberDem0n** [comment]
> Patroni version: 2.1.2

https://github.com/zalando/patroni/pull/2185 fixed it, you have to update to 2.1.3

**Comment by 107142** [comment]
Did that yesterday. Very happy it was found and fixed.

I am still be curious to know why should API endpoint not determine healthiness of the node. If you ever feel like explaining I would appreciate it.

Thanks for the assistance. I shall close it for now.
